intake_esm/cmip.py

```diff
@@ -225,31 +213,71 @@ def results(self):
             self.query_results = get_subset(self.collection_name, self.collection_type, self.query)
             return self.query_results
 
+    def to_xarray(self, **kwargs):
+        """Return dataset as an xarray instance"""
+        _kwargs = self.kwargs.copy()
+        _kwargs.update(kwargs)
+        self.kwargs = _kwargs
+        return self.to_dask()
+
     def _open_dataset(self):
 
-        kwargs = self._kwargs
-        if 'concat_dim' not in kwargs.keys():
-            kwargs.update(concat_dim=self.concat_dim)
-        if self.pattern:
-            kwargs.update(preprocess=self._add_path_to_ds)
+        kwargs = self.kwargs
+        if self.query_results.empty:
+            raise ValueError(f'Query={self.query} returned empty results')
+
+        query = dict(self.query)
+
+        if 'decode_times' not in kwargs.keys():
+            kwargs.update(decode_times=False)
+        if 'time_coord_name' not in kwargs.keys():
+            kwargs.update(time_coord_name='time')
+        if 'ensemble_dim_name' not in kwargs.keys():
+            kwargs.update(ensemble_dim_name='member_id')
+        if 'chunks' not in kwargs.keys():
+            kwargs.update(chunks=None)
 
         # Check that the same variable is not in multiple realms
         realm_list = self.query_results['realm'].unique()
-        if len(realm_list) != 1:
+        frequency_list = self.query_results['frequency'].unique()
+        if len(realm_list) > 1:
             raise ValueError(
-                f"Found in multiple realms:\n \
-                          '\t{realm_list}. Please specify the realm to use"
+                f'Found multiple realms: {realm_list} in query results. Please specify the realm to use'
             )
 
-        ds_dict = OrderedDict()
-        for ens in self.query_results['ensemble'].unique():
-            ens_match = self.query_results['ensemble'] == ens
-            paths = self.query_results.loc[ens_match]['file_fullpath'].tolist()
-            ds_dict[ens] = paths
-
-        ds_list = [
-            xr.open_mfdataset(paths_, chunks=self.chunks, **kwargs) for paths_ in ds_dict.values()
-        ]
-        ens_list = list(ds_dict.keys())
-        self._ds = xr.concat(ds_list, dim='ensemble')
-        self._ds['ensemble'] = ens_list
+        if len(frequency_list) > 1:
+            raise ValueError(
+                f'Found multiple data frequencies: {frequency_list} in query results. Please specify the frequency to use'
+            )
+
+        _ds_dict = {}
+        grouped = get_subset(self.collection_name, self.collection_type, query).groupby(
+            'institution'
+        )
```

Comment:
Does the code above belong in a `_validate_concat_open_dataset` method?

Reply:
Could you expand on this?

Reply:
I think it's probably fine for now. I was wondering whether you could write a general method for validation. Not sure of the API....
```python
def _validate_concat(check_max_instance=None):

  if check_max_instance is not None:
    for fld, max_inst in check_max_instance.items():
        fld_list = self.query_results[fld].unique()
        if len(fld_list) > max_inst:
            raise ValueError(f'message about {fld}')
```

Reply:
I see! Now that we are handling an addition case of  datasets that cannot be concatenated by returning a dictionary of dsets in:

https://github.com/NCAR/intake-esm/blob/eb8fdafa0d7b22770bffde968bf36453d9cd5b9a/intake_esm/cmip.py#L254-L264

can you think of other cases that wouldn't be handled by the above code?

Reply:
I probably need to work thru some use cases to really develop my intuition here.

Reply:
Sounds good! Let me know when you get to play with this.